File: task-browser/src/services/task.service.ts
import http from './http-client'
import {Task} from "../models/task.model";
import {AxiosResponse} from "axios";
import {TaskDTO} from "../models/task-dto.model";

let taskPath = '/task'

class TaskService {
    async getAll(): Promise<AxiosResponse<Array<Task>>> {
        return await http.get(`${taskPath}`);
    }

    async get(id: number): Promise<AxiosResponse<Task>> {
        return await http.get(`${taskPath}/${id}`);
    }

    async create(taskDTO: TaskDTO): Promise<AxiosResponse<Task>> {
        return await http.post(`${taskPath}`, taskDTO);
    }

    async update(id: number, taskDTO: TaskDTO): Promise<AxiosResponse<Task>> {
        return await http.put(`${taskPath}/${id}`, taskDTO);
    }

    async delete(id: number): Promise<AxiosResponse<boolean>> {
        return await http.delete(`${taskPath}/${id}`);
    }

    async findByDescription(description: string): Promise<AxiosResponse<Array<Task>>> {
        return await http.get(`${taskPath}/search?description=${description}`);
    }

}

export default new TaskService();

File: task-service/src/main/kotlin/org/taskmanager/task/service/TaskService.kt
package org.taskmanager.task.service

import kotlinx.coroutines.flow.map
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.taskmanager.task.repository.TaskRepository
import org.taskmanager.task.service.dto.TaskDTO
import org.taskmanager.task.service.mapper.toTask
import org.taskmanager.task.service.mapper.toTaskDTO

@Service
class TaskService(private val taskRepository: TaskRepository) {

    fun findAll() = taskRepository.findAll().map { it.toTaskDTO() }
    suspend fun findById(id: Long) = taskRepository.findById(id)?.toTaskDTO()
    fun findByDescription(description: String) =
        taskRepository.findByDescriptionContainingIgnoreCase(description).map { it.toTaskDTO() }

    fun findByCompleted(completed: Boolean) = taskRepository.findByCompleted(completed).map { it.toTaskDTO() }
    suspend fun create(taskDTO: TaskDTO) = taskRepository.save(taskDTO.toTask()).toTaskDTO()

    @Transactional
    suspend fun update(taskDTO: TaskDTO): TaskDTO? {
        val id = taskDTO.id!!
        if (findById(id) == null) return null
        return taskRepository.save(taskDTO.copy(id = id).toTask()).toTaskDTO()
    }

    @Transactional
    suspend fun delete(id: Long): Boolean {
        if (findById(id) == null) return false
        taskRepository.deleteById(id)
        return true
    }

}
